src/init/ClusterManager.ts

```diff
@@ -0,0 +1,114 @@
+import type { Worker } from 'cluster';
+import cluster from 'cluster';
+import { cpus } from 'os';
+import { getLoggerFor } from '../logging/LogUtil';
+
+/**
+ * Different cluster modes.
+ */
+enum ClusterMode {
+  /** Scales in relation to `core_count`. */
+  autoScale = 0,
+  /** Single threaded mode, no clustering */
+  singleThreaded = 1,
+  /** Fixed amount of workers being forked. (limited to core_count) */
+  fixed = 2
+}
+
+/**
+ * Convert workers amount to {@link ClusterMode}
+ * @param workers - Amount of workers (typically passed through cli parameter)
+ * @returns ClusterMode enum value
+ */
+function toClusterMode(workers: number | string): ClusterMode {
+  if (typeof workers === 'string') {
+    workers = Number.parseInt(workers, 10);
+  }
+  // Return enum from 0 to 2 (inclusive)
+  return Math.max(0, Math.min(2, workers));
+}
+
+/**
+ * This class is responsible for deciding how many affective workers are needed.
+ * It also contains the logic for respawning workers when they are killed by the os.
+ *
+ * The workers values are interpreted as follows:
+ * value | acutal workers |
+ * ------|--------------|
+ * `-1` | `num_cores - 1` workers _(autoscale)_ |
+ * `0` | `num_cores` workers _(autoscale)_ |
+ * `1` | `single threaded mode` _(default)_ |
+ * `2` | `2` workers (but never more than `num_cores`) |
+ * `n` | `n` workers (but never more than `num_cores`) |
+ */
+export class ClusterManager {
+  private readonly logger = getLoggerFor(this);
+  private readonly workers: number;
+  private readonly clusterMode: ClusterMode;
+  private readonly currentWorkers: Worker[] = [];
+  private counter = 0;
+
+  public constructor(workers: number) {
```

Comment:
```suggestion
  public constructor(workers: number | string) {
```
Although it should never be a string though, that's a bug: #1182